2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+These cmdlets will output a warning to use the new cmdlets.
+
+### Local cache
+
+Rather than connecting to PSGallery and other repositories for finding modules,
+the user must explicitly request the current metadata of
+modules from the repositories and that is cached locally.
+`Find-PSResource` (see below) works against this local cache.
+This will also enable changes in PowerShell to use `Find-PSResource -Type Command` to look
+in this cache when it can't find a command and suggest to the user the module to install to
+get that command.
+This will be a local json file containing only the latest version of each module.
+The cache will be stored in the user path.
+There is no system cache that is shared.
+
+Example cache entry:
+
+```json
+{
+  "name": "This is my module",
+  "version": "1.0.0.0",
+  "type": "module",
+  "tags": [
+    "Linux",
+    "PSEdition_Core",
+    "PSEdition_Desktop",
+    "AzureAutomation"
+  ]
+}
+```
+
+An example cache with 5000 resources is ~700KB in compressed json form.
+
+### Automatic updating of the cache
+
+On any operation that reaches out to a repository, a REST API will be called to
+see if the hash of the cache matches the current cache and if not, a new one
+is downloaded.
+If the repository doesn't support this new API, it falls back to current behavior
+in PSGet v2.
+
+### Repository management
+
+`Register-PSResourceRepository` will allow for registering additional repositories.
+A `-Default` switch enables registering PSGallery should it be accidentally removed.
+The `-URL` will accept the HTTP address without the need to specify `/api/v2` as
+that will be assumed and discovered at runtime.
+Support for local filesystem repositories must be maintained.
+A `-InstallationPolicy` switch accepts `Trusted` or `Untrusted` (default) indicating
+whether to prompt the user when installing resources from that repository.
+
+`Get-PSResourceRepository` will list out the registered repositories.
+
+`Set-PSResourceRepository` can be used to update a repository URL or installation policy.
+
+`Unregister-PSResourceRepository` can be used to un-register a repository.
+
+### Finding resources
+
+Current design of PowerShellGet is to have different cmdlets for each of the
+different resources types that it supports:
+
+`Find-Command`
+`Find-DscResource`
+`Find-Module`
+`Find-RoleCapability`
+`Find-Script`
+
+Instead, these will be abstracted by a `Find-PSResource` cmdlet.
+A `-Type` parameter will accept an array of types to allow filtering.
+
+With support of the generic `PSResource`, this means we can also find and
+install arbitrary nupkgs that may only contain an assembly the user wants to
+use in their script.
+
+If the local cache does not exist, the cmdlet will call `Update-PSResourceCache`
+first and then attempt a search.
+However, this cmdlet does not update the cache if one exists.
+
+### Installing resources
+
+`Install-PSResource` will only work for modules.
```

Comment:
If `Install-PSResource` only works for Modules then are you planning on ditching script support, if so why?

Reply:
`Save-PSResource` would work for scripts.  Do people use `Install-Script` where it downloads the script to "~/Documents/PowerShell/Scripts"?

Reply:
I haven't seen it get a lot of use, but I have seen it used on occasion.

Reply:
My organization uses `Install-Script` to deploy installation, monitoring, etc scripts that don't make sense to be in modules. Then we can review installed scripts and versions for compliance. I also like that the location gets added to the path environment variable so scripts can be called without the full path.

Reply:
I've only used `Install-Script` once when it was first available - just to try it out.  I guess I just prefer modules.

Reply:
Install-Script is much more friendly than Save-Script. 

For instance, if the script has module dependencies, Save dumps the modules in the folder you're saving to (even if you already have them installed?) ... which means you then have to deal with (temporarily?) adding that path to PSModulePath, because you can't really manually load the modules  (you don't know their names, versions, or dependency order).

Reply:
@edyoung should have metrics from calls to the current PSGallery that you could use to see the popularity of `Install-Script` and considering there are currently 884 scripts in the gallery this would imply it's reasonably well used.

Reply:
@kilasuit I don't think there's any way the API calls can tell the difference between `Save` and `Install`

Reply:
I believe the latest updates address this.  A script is treated like any other resource.  Modules are only special in that they have a default target destination.